### PR TITLE
Convert entity `type`s to an enum

### PR DESCRIPTION
**Pull request by AllyTally**
## Changes:

In an effort to remove magic numbers, I've given every entity type a name. Hopefully I didn't miss anywhere.

## Legal Stuff:

By submitting this pull request, I confirm that...

- [x] My changes may be used in a future commercial release of VVVVVV
- [x] I will be credited in a `CONTRIBUTORS` file and the "GitHub Friends"
  section of the credits for all of said releases, but will NOT be compensated
  for these changes

### CODE REVIEW

**Comment by InfoTeddy** [comment]
This has conflicts.

Since I can't test this due to conflicts, may I ask what happens if you spawn an invalid entry with `createentity`?

**Comment by AllyTally** [comment]
> may I ask what happens if you spawn an invalid entry with `createentity`?

It spawns the default entity, being the entity which takes player input, does not have a speed cap, uses tile 0, is cyan, etc. Commonly known as a "glitchy Viridian".

**Comment by InfoTeddy** [comment]
So I don't think this was known before, but if you `createentity` 100, that invalid entity is still treated as a teleporter by `entityclass::getteleporter`, and can, for example, fling Viridian out of it with teleporter gamestates.

I don't think any custom levels use this, though.

File: desktop_version/src/Ent.h [hunk 1]
```diff
@@ -5,6 +5,34 @@
 
 #define        rn( rx,  ry) ((rx) + ((ry) * 100))
 
+enum EntityTypes
```
Comment (Daaaav) [inline desktop_version/src/Ent.h:8]: I'd personally call this `EntityType`, so that a variable of that type isn't an "entity types".

**Review by Daaaav — CHANGES_REQUESTED** [review]
Found a couple of places where the enum values aren't used yet:

Script.cpp:
```cpp
                        if (obj.entities[edi].type == 11)
```

LevelDebugger.cpp:
```cpp
            // For gravity lines, show the true hitbox.
            if (obj.entities[i].type == 9)
            {
                graphics.draw_rect(bounding_box.x - 1, bounding_box.y + 1, bounding_box.w + 2, bounding_box.h, graphics.getRGB(90, 90, 15));
            }
            else if (obj.entities[i].type == 10)
            {
                graphics.fill_rect(bounding_box.x - 2, bounding_box.y - 1, bounding_box.w + 1, bounding_box.h + 2, graphics.getRGB(90, 90, 15));
            }
```
```cpp
                // Mostly contains duplicates, but for ease of use
                switch (entity->type)
                {
                case 0:
                    // ...
                case 1:
                    // ...
                case 7:
                    // ...
                case 8:
                    // ...
                case 9:
                    // ...
                case 10:
                    // ...
                }
```
That last `switch` should probably also have a `default:` in it, since my compiler is now warning about missing cases.

Logic.cpp:
```cpp
                if (obj.entities[i].type >= 51
                && obj.entities[i].type <= 54) /* Don't warp warp lines */
```
```cpp
                if ((obj.entities[i].type >= 51
                && obj.entities[i].type <= 54) /* Don't warp warp lines */
```

**Comment by NyakoFox** [comment]
Fixed and force pushed.

**Review by Daaaav — CHANGES_REQUESTED** [review]
One remaining thing:
```
Entity.cpp: In member function ‘bool entityclass::updateentities(int)’:
Entity.cpp:2245:15: warning: enumeration value ‘EntityType_INVALID’ not handled in switch [-Wswitch]
 2245 |         switch(entities[i].type)
      |               ^
```

Otherwise, lgtm! (I changed `entclass.type` to a wrapper struct and manually edited the patch itself to check we're not missing any of them)

**Comment by NyakoFox** [comment]
Fixed that as well!

**Pull request merged**